### ISSUE TITLE
Error importing plugin "pydantic.mypy": No module named 'pydantic'

### ISSUE THREAD

**Issue by slado122** [samuelcolvin/pydantic#3953]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.9.0
            pydantic compiled: True
                 install path: /home/slado122/Programming/Tests/test_pydantic_mypy/.venv/lib/python3.10/site-packages/pydantic
               python version: 3.10.2 (main, Jan 15 2022, 19:56:27) [GCC 11.1.0]
                     platform: Linux-5.16.15-arch1-1-x86_64-with-glibc2.35
     optional deps. installed: ['typing-extensions']
...
```

Pydantic's mypy plugin doesn't work for me. The command to reproduce:
```commandline
python -m venv .venv && source .venv/bin/activate && pip install pydantic && echo "[mypy]\nplugins = pydantic.mypy" > mypy.ini && touch a.py && mypy .
```
Output:
```
Collecting pydantic
  Using cached pydantic-1.9.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.3 MB)
Collecting typing-extensions>=3.7.4.3
  Using cached typing_extensions-4.1.1-py3-none-any.whl (26 kB)
Installing collected packages: typing-extensions, pydantic
Successfully installed pydantic-1.9.0 typing-extensions-4.1.1
WARNING: You are using pip version 21.2.4; however, version 22.0.4 is available.
You should consider upgrading via the '/home/slado122/Programming/Tests/test_pydantic_mypy/.venv/bin/python -m pip install --upgrade pip' command.
mypy.ini:2: error: Error importing plugin "pydantic.mypy": No module named 'pydantic'
Found 1 error in 1 file (errors prevented further checking)
```

**Comment by samuelcolvin** [comment]
Must be a problem with how you have pydantic and/or mypy installed. Works fine for me.

**Comment by mbnoimi** [comment]
I've the same issue occurs with me.
```
# python -c "import pydantic.utils; print(pydantic.utils.version_info())"
             pydantic version: 1.9.1
            pydantic compiled: True
                 install path: /usr/local/lib/python3.9/site-packages/pydantic
               python version: 3.9.7 (default, Sep 28 2021, 18:32:58)  [GCC 10.2.1 20210110]
                     platform: Linux-5.10.104-linuxkit-x86_64-with-glibc2.31
     optional deps. installed: ['dotenv', 'email-validator', 'typing-extensions']
```
**mypy.ini**
```ini
[mypy]
plugins = pydantic.mypy, sqlmypy
ignore_missing_imports = True
disallow_untyped_defs = True
```

```
# poetry show
Skipping virtualenv creation, as specified in config file.
alembic            1.8.1     A database migration tool for SQLAlchemy.
amqp               2.6.1     Low-level AMQP client for Python (fork of amqplib).
anyio              3.6.1     High level compatibility layer for multiple asynchronous event loop implementations
appdirs            1.4.4     A small Python module for determining appropriate platform-specific dirs, e.g. a ...
attrs              21.4.0    Classes Without Boilerplate
autoflake          1.4       Removes unused imports and unused variables
bcrypt             3.2.2     Modern password hashing for your software and your servers
billiard           3.6.4.0   Python multiprocessing fork with improvements and bugfixes
black              19.10b0   The uncompromising code formatter.
cachetools         5.2.0     Extensible memoizing collections and decorators
celery             4.4.7     Distributed Task Queue.
certifi            2022.6.15 Python package for providing Mozilla's CA Bundle.
cffi               1.15.1    Foreign Function Interface for Python calling C code.
chardet            5.0.0     Universal encoding detector for Python 3
charset-normalizer 2.1.0     The Real First Universal Charset Detector. Open, modern and actively maintained a...
click              8.1.3     Composable command line interface toolkit
coverage           6.4.2     Code coverage measurement for Python
cryptography       37.0.4    cryptography is a package which provides cryptographic recipes and primitives to ...
cssselect          1.1.0     cssselect parses CSS3 Selectors and translates them to XPath 1.0
cssutils           2.5.0     A CSS Cascading Style Sheets library for Python
dnspython          2.2.1     DNS toolkit
ecdsa              0.18.0    ECDSA cryptographic signature library (pure python)
email-validator    1.2.1     A robust email syntax and deliverability validation library.
emails             0.6       Modern python library for emails.
fastapi            0.79.0    FastAPI framework, high performance, easy to learn, fast to code, ready for produ...
flake8             3.9.2     the modular source code checker: pep8 pyflakes and co
greenlet           1.1.2     Lightweight in-process concurrent programming
gunicorn           20.1.0    WSGI HTTP Server for UNIX
h11                0.13.0    A pure-Python, bring-your-own-I/O implementation of HTTP/1.1
idna               3.3       Internationalized Domain Names in Applications (IDNA)
iniconfig          1.1.1     iniconfig: brain-dead simple config-ini parsing
isort              4.3.21    A Python utility / library to sort Python imports.
jinja2             2.11.3    A very fast and expressive template engine.
kombu              4.6.11    Messaging library for Python.
lxml               4.9.1     Powerful and Pythonic XML processing library combining libxml2/libxslt with the E...
mako               1.2.1     A super-fast templating language that borrows the best ideas from the existing te...
markupsafe         2.1.1     Safely add untrusted strings to HTML/XML markup.
mccabe             0.6.1     McCabe checker, plugin for flake8
mypy               0.961     Optional static typing for Python
mypy-extensions    0.4.3     Experimental type system extensions for programs checked with the mypy typechecker.
packaging          21.3      Core utilities for Python packages
passlib            1.7.4     comprehensive password hashing framework supporting over 30 schemes
pathspec           0.9.0     Utility library for gitignore style pattern matching of file paths.
pluggy             1.0.0     plugin and hook calling mechanisms for python
poetryup           0.9.0     Update dependencies and bump their version in the pyproject.toml file
premailer          3.10.0    Turns CSS blocks into style attributes
psycopg2-binary    2.9.3     psycopg2 - Python-PostgreSQL Database Adapter
py                 1.11.0    library with cross-python path, ini-parsing, io, code, log facilities
pyasn1             0.4.8     ASN.1 types and codecs
pycodestyle        2.7.0     Python style guide checker
pycparser          2.21      C parser in Python
pydantic           1.9.1     Data validation and settings management using python type hints
pyflakes           2.3.1     passive checker of Python programs
pyparsing          3.0.9     pyparsing module - Classes and methods to define and execute parsing grammars
pytest             7.1.2     pytest: simple powerful testing with Python
pytest-cov         2.12.1    Pytest plugin for measuring coverage.
python-dateutil    2.8.2     Extensions to the standard Python datetime module
python-jose        3.3.0     JOSE implementation in Python
python-multipart   0.0.5     A streaming multipart parser for Python
pytz               2022.1    World timezone definitions, modern and historical
raven              6.10.0    Raven is a client for Sentry (https://getsentry.com)
regex              2022.7.9  Alternative regular expression module, to replace re.
requests           2.28.1    Python HTTP for Humans.
rsa                4.8       Pure-Python RSA implementation
six                1.16.0    Python 2 and 3 compatibility utilities
sniffio            1.2.0     Sniff out which async library your code is running under
sqlalchemy         1.4.39    Database Abstraction Library
sqlalchemy-stubs   0.3       SQLAlchemy stubs and mypy plugin
starlette          0.19.1    The little ASGI library that shines.
tenacity           6.3.1     Retry code until it succeeds
toml               0.10.2    Python Library for Tom's Obvious, Minimal Language
tomli              2.0.1     A lil' TOML parser
tomlkit            0.11.1    Style preserving TOML library
typed-ast          1.5.4     a fork of Python 2 and 3 ast modules with type comment support
typer              0.4.2     Typer, build great CLIs. Easy to code. Based on Python type hints.
typing-extensions  4.3.0     Backported and Experimental Type Hints for Python 3.7+
urllib3            1.26.10   HTTP library with thread-safe connection pooling, file post, and more.
uvicorn            0.18.2    The lightning-fast ASGI server.
vine               1.3.0     Promises, promises, promises.
```

**Comment by mbnoimi** [comment]
> 



> Must be a problem with how you have pydantic and/or mypy installed. Works fine for me.

For me. I installed both of them using `poetry`
**pyproject.toml**
```toml
[tool.poetry]
name = "app"
version = "0.1.0"
description = "A Personal Todo app"
authors = ["Admin <admin@bayti.top>"]

[tool.poetry.dependencies]
python = "^3.9"
uvicorn = "^0.18.2"
fastapi = "^0.79.0"
python-multipart = "^0.0.5"
email-validator = "^1.0.5"
requests = "^2.28.1"
celery = "^4.4.2"
passlib = {extras = ["bcrypt"], version = "^1.7.2"}
tenacity = "^6.1.0"
pydantic = "^1.9.1"
emails = "^0.6"
raven = "^6.10.0"
gunicorn = "^20.0.4"
jinja2 = "^2.11.2"
psycopg2-binary = "^2.8.5"
alembic = "^1.4.2"
pytest = "^7.1.2"
python-jose = {extras = ["cryptography"], version = "^3.1.0"}
SQLAlchemy = "^1.4.39"

[tool.poetry.dev-dependencies]
mypy = "^0.961"
black = "^19.10b0"
isort = "^4.3.21"
autoflake = "^1.3.1"
flake8 = "^3.7.9"
pytest = "^7.1.2"
sqlalchemy-stubs = "^0.3"
pytest-cov = "^2.8.1"
poetryup = "^0.9.0"

[tool.isort]
multi_line_output = 3
include_trailing_comma = true
force_grid_wrap = 0
line_length = 88
[build-system]
requires = ["poetry>=0.12"]
build-backend = "poetry.masonry.api"
```

**Comment by mbnoimi** [comment]
> 
> Must be a problem with how you have pydantic and/or mypy installed. Works fine for me.

For me. I installed both of them using `poetry`
**pyproject.toml**
```toml
[tool.poetry]
name = "app"
version = "0.1.0"
description = "A Personal Todo app"
authors = ["Admin <admin@expmaple.com>"]

[tool.poetry.dependencies]
python = "^3.9"
uvicorn = "^0.18.2"
fastapi = "^0.79.0"
python-multipart = "^0.0.5"
email-validator = "^1.0.5"
requests = "^2.28.1"
celery = "^4.4.2"
passlib = {extras = ["bcrypt"], version = "^1.7.2"}
tenacity = "^6.1.0"
pydantic = "^1.9.1"
emails = "^0.6"
raven = "^6.10.0"
gunicorn = "^20.0.4"
jinja2 = "^2.11.2"
psycopg2-binary = "^2.8.5"
alembic = "^1.4.2"
pytest = "^7.1.2"
python-jose = {extras = ["cryptography"], version = "^3.1.0"}
SQLAlchemy = "^1.4.39"

[tool.poetry.dev-dependencies]
mypy = "^0.961"
black = "^19.10b0"
isort = "^4.3.21"
autoflake = "^1.3.1"
flake8 = "^3.7.9"
pytest = "^7.1.2"
sqlalchemy-stubs = "^0.3"
pytest-cov = "^2.8.1"
poetryup = "^0.9.0"

[tool.isort]
multi_line_output = 3
include_trailing_comma = true
force_grid_wrap = 0
line_length = 88
[build-system]
requires = ["poetry>=0.12"]
build-backend = "poetry.masonry.api"
```

**Comment by nerewarin** [comment]
why it's closed? problem is actual

**Comment by samuelcolvin** [comment]
To reopen and/or help I need some more information on your set up and what you've tried to solve this.

**Comment by DbCrWk** [comment]
> why it's closed? problem is actual

@nerewarin I had this issue with my pyproject.toml set to:
```
mypy = "^0.971"
pydantic = "^1.9.2"
```

but by updating to
```
pydantic = "^1.10.2"
mypy = "^0.982"
```

I was able to get past this error.

**Comment by DbCrWk** [comment]
Oddly, if I use exact versions, the issue goes away. I'm guessing there was some very specific transient issue with `mypy` and `pydantic` that is now resolved.

**Comment by cahna** [comment]
I encountered [this same issue](https://github.com/cahna/hitfactorpy/actions/runs/3973249231/jobs/6811785290).

@DbCrWk's advice worked locally, but doesn't seem to work on github actions. I'm still investigating.

**Comment by Jylpah** [comment]
I am having the very same problem with mypy 1.2.0 and pydantic 1.10.7. Could someone post a complete solution. The issue is closed, but the solution is unclear.

**Comment by dmontagu** [comment]
In many of the cases above, it looks like it's possible it was using the `mypy` installed in a _different_ mypy environment. Of course I can't confirm that is the case, but if you are running `pip install pydantic` or similar, I would suggest making sure that the full paths returned by `which pip` and `which mypy` are the same — you need to make sure the mypy that is used for type checking is in an environment where pydantic is available.

To check if it _should_ be working, you might be able to run something like this:
```
"$(dirname $(which mypy))"/python -c 'import pydantic' && echo 'pydantic is installed'
# If successful: pydantic is installed
# If unsuccessful: ModuleNotFoundError: No module named 'pydantic'
```
In principle you should be able to install pydantic into mypy's environment via:
```
"$(dirname $(which mypy))"/pip install pydantic==<version>
```
but rather than doing the above, I would suggest figuring out why mypy is in a different environment and fixing that issue. (Perhaps you need to be running `pip install mypy` or similar)

**Comment by Jylpah** [comment]
I am running `mypy` (`dmypy`) using VS Code's Mypy-plugin using virtual environment under the repo. I have enabled an option to run `mypy` using active interpreter. 

VS Code's `dmypy` output looks like this: 

```
[10] stderr:
pyproject.toml:1:1: error: Error importing plugin "pydantic.mypy": No module named 'pydantic'  [misc]

[10] Error running mypy in /home/jylpah/src/pyutils: mypy failed with error: "pyproject.toml:1:1: error: Error importing plugin "pydantic.mypy": No module named 'pydantic'  [misc]". See Output panel for details.
Mypy settings changed: /home/jylpah/src/pyutils
[11] Check folder: /home/jylpah/src/pyutils
[11] Received python path from Python extension: /home/jylpah/src/pyutils/.venv/bin/python
[11] Running dmypy in folder /home/jylpah/src/pyutils
/home/jylpah/src/pyutils/.venv/bin/python -m mypy.dmypy --status-file /home/jylpah/.vscode-server/data/User/workspaceStorage/2fdff59c5118f1fdcf86eb0ea850ff65/matangover.mypy/dmypy-87d788e607b40726b815ff05bafaf021d67bd81d-187272.json run --log-file /home/jylpah/.vscode-server/data/User/workspaceStorage/2fdff59c5118f1fdcf86eb0ea850ff65/matangover.mypy/dmypy-87d788e607b40726b815ff05bafaf021d67bd81d.log -- . --show-column-numbers --no-error-summary --no-pretty --no-color-output --python-executable /home/jylpah/src/pyutils/.venv/bin/python
[11] stderr:
pyproject.toml:1:1: error: Error importing plugin "pydantic.mypy": No module named 'pydantic'  [misc]

[11] Error running mypy in /home/jylpah/src/pyutils: mypy failed with error: "pyproject.toml:1:1: error: Error importing plugin "pydantic.mypy": No module named 'pydantic'  [misc]". See Output panel for details.
```


```
% pip freeze
...
mypy==1.2.0
mypy-extensions==1.0.0
...
pydantic==1.10.7
...
```

**Comment by bfontaine** [comment]
I had a a similar issue as "`Error importing plugin "pydantic.mypy": No module named 'pydantic_core'`". I tried hard to find the root cause, but at the end the only thing that "fixed" it was to delete the virtualenv; remove the private packages and `tool.poetry.source` from `pyproject.toml`; `poetry install`; then put them back and `poetry install` again.

This is the minimal `pyproject.toml` that reproduce the issue with Poetry 1.7.1, although you need a valid PyPI server URL instead of `https://pypi.example.com/simple/`:
```toml
[tool.poetry]
name = "foo"
version = "0.1.0"
description = ""
authors = []
readme = "README.md"

[tool.poetry.dependencies]
python = "3.10.14"
pydantic = "2.7.1"
mypy = "1.10.0"

[[tool.poetry.source]]
name = "x"
url = "https://pypi.example.com/simple/"
priority = "default"

[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"

[tool.mypy]
plugins = ["pydantic.mypy"]
```
If I run `poetry install` with it, I get the issue. If I remove the virtualenv and the `tool.poetry.source`, and then `poetry install`, it works fine.

This is definitely not an issue with Pydantic itself.

**Comment by woile** [comment]
I recently started having this issue, I had only `pydantic-settings` and things were working fine. Last week `mypy` (with `plugins = ["pydantic.mypy"]`) started complaining that `pydantic-core` is missing, even after installing it, it complained that pydantic cannot find `annotated-types`.

I've tried adding just `pydantic` but it doesn't seem to include `pydantic-core` nor `annotated-types`, the strange thing is that I see them as dependencies: https://github.com/pydantic/pydantic/blob/main/pyproject.toml#L48-L52

**Comment by vervoorn** [comment]
I have the same issue

**Comment by Casyfill** [comment]
fwiw, I had this problem when bumped Pydantic from v1 to v2; I hoped to not do anything with the code itself (tests are passing) but seems that trick didnt work with mypy: had to set plugin to v1 to work:
```
[mypy]
plugins = pydantic.v1.mypy
```

**Comment by Andriy-Sydorenko** [comment]
> I encountered [this same issue](https://github.com/cahna/hitfactorpy/actions/runs/3973249231/jobs/6811785290).
> 
> @DbCrWk's advice worked. ~but doesn't seem to work on github actions. I'm still investigating...~
> 
> Update: The github actions issue was a user error related to `pre-commit`; I needed to add an `additional_dependencies` section to my `mypy` config in `.pre-commit-config.yaml`:
> 
> ```yaml
> repos:
>  - repo: https://github.com/pre-commit/mirrors-mypy
>    rev: v0.991
>    hooks:
>      - id: mypy
>        files: ^(mypackage/|tests/)
>        additional_dependencies:
>         - "pydantic>=1.10.4"
> ```

man, you're my savior

**Comment by mhadam** [comment]
Just in case anyone ends up here, mine turned out to be me failing to understand the difference between `uvx mypy ...` and `uv run mypy ...`.

`uvx` runs tools independently from your project, it's _not_ a shorthand for `uv run`.

tl;dr: use `uv run mypy ...` instead.

**Comment by asibs** [comment]
I'm experiencing this issue when using the experimental no-gil mode on Python 3.13 & poetry.

**Minimal reproduction:**

(using `pyenv` & `poetry`)

1. `mkdir mypy-test && cd mypy-test`
1. Build python 3.13 with the `--disable-gil` flag, and store in your pyenv versions folder: `PYTHON_CONFIGURE_OPTS='--disable-gil' python-build 3.13.2 /Users/my.name/.pyenv/versions/3.13.2_nogil`
1. Use the no-gil version of python in the current folder: `pyenv local 3.13.2_nogil`
1. Test the GIL is disabled: `poetry run python -c "import sys; print(sys._is_gil_enabled());"` (should print `False`)
1. Create a minimal `pyproject.toml`:

```
[tool.poetry]
name = "mypy-test"
version = "0.1.0"
description = "Minimal reproduction of mypy issue with Pydantic on Python 3.13"
package-mode = false

[tool.poetry.dependencies]
python = "^3.13"
pydantic = "^2.10"

[tool.poetry.group.dev.dependencies]
mypy = "^1.15.0"

[tool.mypy]
plugins = ["pydantic.mypy"]
```

1. Install dependencies (Pydantic / MyPy): `poetry install`
1. Create a hello world Python file in the project directory so mypy has something to do
1. Run mypy via poetry: `poetry run mypy -v .`

Output is:

```
pyproject.toml:1: error: Error importing plugin "pydantic.mypy": No module named 'pydantic_core._pydantic_core'  [misc]
Found 1 error in 1 file (errors prevented further checking)
```

If you install regular Python 3.13.2 (without the `--disable-gil` flag), this error doesn't happen - mypy runs successfully.

**Versions I'm using:**
- poetry = 1.8.3
- pyenv = 2.5.3
- python = 3.13.2 (built with `--disable-gil` flag, see instructions above)
- pydantic = 2.10.6
- mypy = 1.15.0

**Comment by bowenliang123** [comment]
I've got similar issue with Pydantic 2.11 and Pydantic-settings 2.8.1 on UV tool and Python 3.12.

```
mypy.ini:10: error: Error importing plugin "['pydantic.mypy']": No module named "['pydantic"  [misc]
```

**Comment by umutseven92** [comment]
Having the same issue as @asibs; `Error importing plugin` error coming from Pydantic when running project with Python 3.13.4, built with the `--disable-gil` flag.

**Comment by Viicos** [comment]
I'm locking this issue since this isn't going anywhere. To sum things up:

- Mypy allows plugins to be used, and can be [declared](https://docs.pydantic.dev/latest/integrations/mypy/#enabling-the-plugin) in the `mypy.ini` or `pyproject.toml` file. The plugin can be declared using the module name (in our case `pydantic.mypy`). Internally, Mypy will try to dynamically import the `pydantic.mypy`, so it is important that **mypy and pydantic are installed in the same environment.**
- If you get the `No module named 'pydantic_core'` error, it either means:
  - You are installing pydantic on an unsupported architecture (unlikely to happen as we build for many targets, the list is available [here](https://pypi.org/project/pydantic_core/#files)).
  - You are installing pydantic on a no-GIL/free-threaded Python build. no-GIL is only (experimentally) supported starting with 2.11, so you will have to upgrade to this version.
- If you are using `uv` (and especially `uvx`), please read [their documentation](https://docs.astral.sh/uv/guides/tools/#using-tools) to understand the different between `uvx` and `uv run`. TL;DR: **don't rely on uvx to run mypy** (it will create a temporary venv where `pydantic` isn't installed).
- If you are using pre-commit: it will create a temporary venv, where pydantic isn't installed. Either use [`language: system`](https://pre-commit.com/#system), or [check this comment](https://github.com/pydantic/pydantic/issues/3953#issuecomment-1399178132).

If things still don't work on your end _after_ checking all the above, please open a new issue with clear reproduction steps (creation of the venv, commands run).
